R/sensitivity_labels.R

```diff
@@ -143,19 +162,105 @@ apply_sensitivity_label <- function(file, label) {
+  }
+
+  xml_map <- .get_sensitivity_xml_map()
+  xml <- xml_map[[label]]
+
+
+  ## Apply label to Excel workbooks ----
+
+  if(file_ext %in% c("xlsx", "xls")){
+    wb <- openxlsx2::wb_load(file)
+    wb <- openxlsx2::wb_add_mips(wb, xml)
+    openxlsx2::wb_save(wb, file)
   }
 
-  # Load workbook and apply label
-  wb <- openxlsx2::wb_load(file)
-  xml <- sensitivity_label_xml[[label]]
-  wb <- openxlsx2::wb_add_mips(wb, xml)
-  openxlsx2::wb_save(wb, file)
+  ## Apply label to Word docs ----
+
+  if(file_ext == "docx"){
+    # Parsing input
+    file_path <- dirname(file)
+    file_name <- basename(file)
+    file_type <- tools::file_ext(file)
+    # Removing file type from actual name
+    file_name <- sub(paste0(".", file_type), "", file_name)
+
+    # Zipping process needs its own directory
+    # We'll reset it using on.exit later
+    current_wd <- getwd()
+
+    zipdir <- paste0(file_path, "/", file_name)
+
+    # Create the dir using that name
+    dir.create(zipdir)
+
+    # Unzip the file into the dir
+    unzip(file, exdir= zipdir)
+
+    # Formatting as xml, not character data.
+    xml <- xml2::as_xml_document(xml)
+
+    # Create the dir using that name
+    dir.create(paste0(zipdir, "/docMetadata"), showWarnings = FALSE, recursive = TRUE)
+
+    # Write the XML data to the temp directory
+    xml2::write_xml(xml, paste0(zipdir, "/docMetadata/LabelInfo.xml"), useBytes = TRUE)
+
+    # Update content file with new child node
+    content <- xml2::read_xml(paste0(zipdir, "/[Content_Types].xml"))
+    new_node <- xml2::xml_add_child(content, .value = "Override")
+
+    xml2::xml_set_attrs(new_node, c(
+      PartName = "/docMetadata/LabelInfo.xml",
+      ContentType = "application/vnd.ms-office.classificationlabels+xml"
+    ))
+
+    xml2::write_xml(content, paste0(zipdir, "/[Content_Types].xml"), useBytes = TRUE)
+
+    # Update .rels file with new child node
+    rels_file <- xml2::read_xml(paste0(zipdir, "/_rels/.rels"))
+
+    xml2::xml_set_attrs(xml2::xml_add_child(rels_file, .value = "Relationship"), c(
+      Id = "rId6",
+      Type="http://schemas.microsoft.com/office/2020/02/relationships/classificationlabels",
+      Target="docMetadata/LabelInfo.xml"
+    ))
+
+
+    xml2::write_xml(rels_file, paste0(zipdir, "/_rels/.rels"), useBytes = TRUE)
+
+    # Delete original file
+    file.remove(file)
+
+    newzip <- paste0(file_path, "/", file_name, ".zip")
+    file.create(newzip)
+
+    setwd(zipdir) # Setting new base directory to avoid needless recursion in file zipping
+    # will be reset on.exit() anyway so don't change.
+
+    suppressWarnings(
+      zip::zip(zipfile = newzip,
+               files = list.files(zipdir, all.files = TRUE, full.names = FALSE, recursive = TRUE, include.dirs = TRUE),
+               recurse = FALSE,
+               include_directories = FALSE,
+               root = ".") # end of zip()
+    ) # End of suppressWarnings
+
```

Comment:
would we not need to set working directory back at this point, i.e. `setwd(current_wd)`?